Fix: LiteLLM Streaming Content Duplication in Tool Call Responses

thesynapses opened the pull request:
# Fix: LiteLLM Streaming Content Duplication in Tool Call Responses

## Summary
Fixes content duplication in streaming responses when using LiteLLM models with ADK's planning features (e.g., `PlanReActPlanner`). Planning and reasoning text was appearing twice: once during streaming as individual chunks, and again in the aggregated tool-call message.

Resolves #3697

## Problem
When the model generates planning/reasoning text (e.g., `<PLANNING>I need to search...</PLANNING>`) followed by tool calls during streaming:

1. **Text chunks are streamed** to users in real-time (lines 1288-1296) 
2. **Same text included again** in aggregated tool-call message with `content=text` (line 1352) 

This violates OpenAI/LiteLLM conventions where tool-call-only messages should have `content=None`.

## Solution
Changed line 1352 to set `content=None` for tool-call messages:

```python
ChatCompletionAssistantMessage(
    role="assistant",
    content=None,  # FIX: Avoid duplication, follow OpenAI spec
    tool_calls=tool_calls,
)
```

## Why This Works
- **Planning text already streamed**: Users see it in real-time via individual chunks
- **Preserved in `thought_parts`**: Reasoning captured separately at line 1357 for conversation history
- **Follows API standards**: OpenAI, Claude, GPT APIs expect `content=None` for tool-only messages
- **Correct semantics**: Tool-call messages represent function invocations, not answer text

## Impact
- Eliminates content duplication
- Aligns with OpenAI/LiteLLM conventions
- Preserves reasoning context via `thought_parts`
- Clean conversation history without redundant content
- Proper semantic representation of tool-call turns

## Testing
Tested with:
- LiteLLM models (Claude, GPT) with planning workflows
- Streaming enabled with tool calls
- Multi-turn conversations requiring tool usage

## Additional Context
- Affects only streaming mode with tool calls
- Non-streaming path already handles this correctly
- `thought_parts` parameter preserves reasoning separately from message content
- No breaking changes to existing APIs

thesynapses committed:
Fix: Map finish_reason for LiteLLM streaming responses

Fixes #3665

Streaming responses from LiteLLM models (Claude, GPT, etc.) were not setting
finish_reason on aggregated LlmResponse objects, causing agent runners to not
properly recognize completion states.

This fix mirrors the finish_reason mapping logic from the non-streaming path
(lines 776-784) and applies it to both streaming code paths:
- Tool call responses (lines 1340-1368)
- Text-only responses (lines 1369-1390)

Without this fix, agents using Claude or GPT via LiteLLM would encounter stop
conditions that couldn't be properly handled, leading to incomplete responses
or unexpected agent behavior.

Tested with Claude Sonnet 4.5 and GPT-5 via Azure OpenAI in production
multi-agent system with MCP tools.

thesynapses committed:
fix: Set content=None for tool-only messages in streaming to avoid duplication

- Fixes content duplication where planning/reasoning text appears twice
  (once during streaming, again in aggregated tool-call message)
- Aligns with OpenAI/LiteLLM conventions for tool-call messages
- Planning text is preserved in thought_parts and already streamed to users
- Resolves semantic confusion where tool-call messages contained text content

Fixes #3697

thesynapses committed:
refactor: Extract _set_finish_reason helper to eliminate code duplication

Extract finish_reason mapping logic into a reusable helper function to address
code duplication feedback from Gemini Code Assist review on PR #3698.

Changes:
- Added _set_finish_reason(response, finish_reason) helper function
- Replaced three duplicate mapping blocks with single helper call:
  * Non-streaming path (line ~880)
  * Streaming tool-call path (line ~1387)
  * Streaming text-only path (line ~1409)
- Preserved all existing comments and behavior
- Improved maintainability - single source of truth for mapping logic

Addresses: https://github.com/google/adk-python/pull/3698\#discussion_r18xxxxx

ryanaiagent commented:
Hi @thesynapses , Thank you for your work on this pull request. We appreciate the effort you've invested.
Before we can proceed with the review can you please fix the lint errors. You can use autoformat.sh.

ryanaiagent commented:
Hi @thesynapses , This PR has merge conflicts that require changes from your end. Could you please rebase your branch with the latest main branch to address these? Once this is complete, please let us know so we can proceed with the review.

thesynapses committed:
Merge remote-tracking branch 'upstream/main' into fix/litellm-streaming-content-duplication

thesynapses committed:
Merge branch 'main' into fix/litellm-streaming-content-duplication

thesynapses committed:
Update lite_llm.py

thesynapses committed:
Merge branch 'fix/litellm-streaming-content-duplication' of https://github.com/thesynapses/adk-python into fix/litellm-streaming-content-duplication

thesynapses committed:
Merge branch 'main' into fix/litellm-streaming-content-duplication

ryanaiagent commented:
Hi @thesynapses , can you also fix the formatting errors and update the branch to main.

thesynapses committed:
fix: Apply pyink formatting and isort to fix CI checks

thesynapses committed:
Merge remote-tracking branch 'upstream/main' into fix/litellm-streaming-content-duplication

rohityan commented:
Hi @thesynapses , Since the requested changes are still pending, I'm going to close this PR. Feel free to reopen once you've had a chance to address the requested changes, or create a new PR. Thanks!

thesynapses commented:
@rohityan - I am not quite clear on what is expected from my end as a contributor:
- I found an issue and it was reported
- Then I fixed the issue
- Then I fixed other ancillary formatting also
- Now you are saying the request is pending. What is missing? We can see that I applied the requested fixes. 

The issues are getting lost in paperwork.
